examples/inspector_with_lifetime/src/main.rs

```diff
@@ -0,0 +1,83 @@
+use std::convert::Infallible;
+
+use revm::{
+    context::{
+        result::{EVMError, ExecutionResult},
+        BlockEnv, CfgEnv, TxEnv,
+    },
+    database::State,
+    inspector::NoOpInspector,
+    primitives::B256,
+    Database, DatabaseCommit, ExecuteCommitEvm, Inspector, Journal, JournalEntry, MainBuilder as _,
+};
+
+pub type ContextRef<'db> = revm::Context<
+    BlockEnv,
+    TxEnv,
+    CfgEnv,
+    &'db mut dyn Database<Error = Infallible>,
+    Journal<&'db mut dyn Database<Error = Infallible>>,
+>;
+
+pub trait DBSuperTrait: Database<Error = Infallible> + DatabaseCommit {}
+
+impl<T: Database<Error = Infallible> + DatabaseCommit> DBSuperTrait for T {}
+
+fn mine_block<InspectorT>(
+    cfg: &CfgEnv,
+    db: &mut dyn DBSuperTrait,
+    transactions: Vec<TxEnv>,
+    inspector: &mut InspectorT,
+) -> Result<Vec<ExecutionResult>, EVMError<Infallible>>
+where
+    InspectorT: for<'context> Inspector<Context<'context> = ContextRef<'context>>,
+{
+    let block = BlockEnv {
+        prevrandao: Some(B256::random()),
+        ..BlockEnv::default()
+    };
+
+    let mut results = Vec::new();
+    for tx in transactions {
+        let result = run_transaction(&block, cfg, db, tx, inspector)?;
+
+        results.push(result);
+    }
+
+    Ok(results)
+}
+
+fn run_transaction<InspectorT>(
+    block: &BlockEnv,
+    cfg: &CfgEnv,
+    db: &mut dyn DBSuperTrait,
+    tx: TxEnv,
+    inspector: &mut InspectorT,
```

Comment:
Yeah, the Inspector takes the lifetime of the Context and it is impossible to use in the way we expect and it sucks. This was a reason why inspector field is part Evm struct, if you send full Evm here it will work, or even impl `InspectEvm`

Reply:
For some of our use cases that's possible, like building a block.

In other cases we construct the EVM ad-hoc. The flow is similar to this, e.g. for `eth_call`:

- Retrieve `state` at the requested block number from `blockchain`
- Construct a `BlockEnv` for the requested block number
- Construct a transaction for the `eth_call` request
- Construct EVM with DatabaseComponents (for `blockchain` and `state`), the requested `eth_call` transaction, `BlockEnv`, and `CfgEnv`
- Run `replay_transact`

I might just have to inline the contruction of an EVM in that case using an efficient builder pattern API.